File: library/src/test/java/me/proxer/library/api/info/ForumDiscussionsEndpointTest.java
package me.proxer.library.api.info;

import me.proxer.library.ProxerTest;
import me.proxer.library.api.ProxerException;
import me.proxer.library.entity.info.ForumDiscussion;
import okhttp3.mockwebserver.MockResponse;
import org.junit.Test;

import java.io.IOException;
import java.util.Date;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author Ruben Gees
 */
public class ForumDiscussionsEndpointTest extends ProxerTest {

    @Test
    public void testDefault() throws IOException, ProxerException {
        server.enqueue(new MockResponse().setBody(fromResource("info_forum.json")));

        final List<ForumDiscussion> result = api.info()
                .forumDiscussions("1")
                .build()
                .execute();

        assertThat(result).first().isEqualTo(buildFirstTestForumDiscussion());
        assertThat(result).last().isEqualTo(buildLastTestForumDiscussion());
    }

    @Test
    public void testPath() throws ProxerException, IOException, InterruptedException {
        server.enqueue(new MockResponse().setBody(fromResource("info_forum.json")));

        api.info().forumDiscussions("3")
                .build()
                .execute();

        assertThat(server.takeRequest().getPath()).isEqualTo("/api/v1/info/forum?id=3");
    }

    private ForumDiscussion buildFirstTestForumDiscussion() {
        return new ForumDiscussion("384098", "281", "Airing-Anime",
                "Overlord II - Diskussionsthread",
                15, 749, new Date(1514199320 * 1000L), "351626", "Asuka..",
                new Date(1517246199 * 1000L), "506979", "5devilz");
    }

    private ForumDiscussion buildLastTestForumDiscussion() {
        return new ForumDiscussion("381421", "56", "Anime- und Manga-News",
                "Overlord II – neues Visual, weitere Charaktere und Synchronsprecher enthüllt",
                35, 32334, new Date(1489228544 * 1000L), "19328", "Moeface",
                new Date(1514078674 * 1000L), "470614", "..Rhyanon.");
    }
}

File: library/src/test/kotlin/me/proxer/library/api/info/ForumDiscussionsEndpointTest.kt
package me.proxer.library.api.info

import me.proxer.library.ProxerTest
import me.proxer.library.entity.info.ForumDiscussion
import me.proxer.library.runRequest
import org.amshove.kluent.shouldEqual
import org.junit.jupiter.api.Test
import java.util.Date

/**
 * @author Ruben Gees
 */
class ForumDiscussionsEndpointTest : ProxerTest() {

    private val expectedFirstForumDiscussion = ForumDiscussion(
        id = "384098", categoryId = "281", categoryName = "Airing-Anime", subject = "Overlord II - Diskussionsthread",
        postAmount = 15, hits = 749, firstPostDate = Date(1514199320 * 1000L), firstPostUserId = "351626",
        firstPostUsername = "Asuka..", lastPostDate = Date(1517246199 * 1000L), lastPostUserId = "506979",
        lastPostUsername = "5devilz"
    )

    private val expectedLastForumDiscussion = ForumDiscussion(
        id = "381421", categoryId = "56", categoryName = "Anime- und Manga-News",
        subject = "Overlord II – neues Visual, weitere Charaktere und Synchronsprecher enthüllt", postAmount = 35,
        hits = 32334, firstPostDate = Date(1489228544 * 1000L), firstPostUserId = "19328",
        firstPostUsername = "Moeface", lastPostDate = Date(1514078674 * 1000L), lastPostUserId = "470614",
        lastPostUsername = "..Rhyanon."
    )

    @Test
    fun testDefault() {
        val (result, _) = server.runRequest("info_forum.json") {
            api.info
                .forumDiscussions("1")
                .build()
                .safeExecute()
        }

        result.first() shouldEqual expectedFirstForumDiscussion
        result.last() shouldEqual expectedLastForumDiscussion
    }

    @Test
    fun testPath() {
        val (_, request) = server.runRequest("info_forum.json") {
            api.info.forumDiscussions("3")
                .build()
                .execute()
        }

        request.path shouldEqual "/api/v1/info/forum?id=3"
    }
}
